ledger/apply/heartbeat.go

```diff
@@ -0,0 +1,102 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package apply
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+)
+
+// Heartbeat applies a Heartbeat transaction using the Balances interface.
+func Heartbeat(hb transactions.HeartbeatTxnFields, header transactions.Header, balances Balances, provider hdrProvider, round basics.Round) error {
+	// Get the account's balance entry
+	account, err := balances.Get(hb.HbAddress, false)
+	if err != nil {
+		return err
+	}
+
+	// In txnGroupBatchPrep, we do not charge for singleton (Group.IsZero)
+	// heartbeats. But we only _want_ to allow free heartbeats if the account is
+	// under challenge. If this is an underpaid singleton heartbeat, reject it
+	// unless the account is under challenge.
+
+	proto := balances.ConsensusParams()
+	if header.Fee.Raw < proto.MinTxnFee && header.Group.IsZero() {
+		kind := "free"
+		if header.Fee.Raw > 0 {
+			kind = "cheap"
+		}
+
+		if account.Status != basics.Online {
+			return fmt.Errorf("%s heartbeat is not allowed for %s %+v", kind, account.Status, hb.HbAddress)
+		}
+		if !account.IncentiveEligible {
+			return fmt.Errorf("%s heartbeat is not allowed when not IncentiveEligible %+v", kind, hb.HbAddress)
+		}
+		ch := FindChallenge(proto.Payouts, round, provider, ChRisky)
+		if ch.IsZero() {
+			return fmt.Errorf("%s heartbeat for %s is not allowed with no challenge", kind, hb.HbAddress)
+		}
+		if !ch.Failed(hb.HbAddress, account.LastSeen()) {
+			return fmt.Errorf("%s heartbeat for %s is not challenged by %+v", kind, hb.HbAddress, ch)
+		}
+	}
+
+	// Note the contrast with agreement. We require the account's _current_
+	// partkey be used to sign the heartbeat. This is required because we can
+	// only look 320 rounds back for voting information. If a heartbeat was
+	// delayed a few rounds (even 1), we could not ask "what partkey was in
+	// effect at firstValid-320?"  Using the current keys means that an account
+	// that changes keys would invalidate any heartbeats it has already sent out
+	// (that haven't been evaluated yet). Maybe more importantly, after going
+	// offline, an account can no longer heartbeat, since it has no _current_
+	// keys. Yet it is still expected to vote for 320 rounds.  Therefore,
+	// challenges do not apply to accounts that are offline (even if they should
+	// still be voting).
+
+	// heartbeats sign a message consisting of the BlockSeed of the first-valid
+	// round, to discourage unsavory behaviour like presigning a bunch of
+	// heartbeats for later use keeping an unavailable account online.
+	hdr, err := provider.BlockHdr(header.FirstValid)
```

Comment:
It seems weird that the header for the transaction's `FirstValid` must already be available in order to process the transaction.  It seems like this means a heartbeat transaction can't really ever be in its `FirstValid` round, because that round must have already been committed (without this transaction) in order to construct this transaction.

Reply:
It does seem weird. But I think it's for the best. It is at least plausible that a node could have committed round `n`, but be talking to a relay that has not. So, I've seen "best practices" recommendations that say that you should not create transactions with FirstValid set to your own `Latest()+1`, but rather `Latest()` so it would not have it's txn dropped by the relay (this seems even more plausible in a future p2p world).  Signing the FirstValid block forces that behavior. These transactions are being autogenerated, so I wanted to make them as bulletproof as possible - a user can't say "oops, didn't work, I'll try again with an older `FirstValid`.

Also, it made the implementation simpler. The service didn't need to look up an old header after it already looked up the most recent header to get the current protocol. But I'll insist I did it for the first reason.